### ISSUE TITLE
Emacs can't activate the input method 'Lean'

### ISSUE THREAD

**Issue by la-marc** [leanprover/lean4-mode#30]
Hi!

I have recently installed lean4-mode, following the instructions. I have used the first, "basic" method (i.e. without use-package).
I am, however, getting an error saying that Emacs can't activate the input method 'Lean'.
This error appears both on my Fedora desktop, as well as on my WSL2 Fedora setup on Windows.
Other Emacs input methods seem to work, e.g. the "greek" input method etc.
Due to this error, the lean4-mode does not start. However, after commenting out
(require 'lean4-input) and (set-input-method "Lean") in lean4-mode.el, it works (but \a does not yield alpha and so on, as expected), so "most" of the installation went as planned.

Any idea, how to amend this?

**Comment by akirak** [comment]
Hello @la-marc ,

This might be due to an update at #23. Do you have `abbreviations.json` in the same directory as `lean4-mode.el`? If not, are you using the following MELPA recipe (or its equivalent for your package manager, e.g. straight.el)?

```elisp
(lean4-mode :fetcher github :repo "leanprover/lean4-mode"
 :files ("*.el" "data")
```

**Comment by la-marc** [comment]
My `abbreviations.json` is in a subdirectory `/data/`, and `/data/` is in the same same directory as `lean4-mode.el`.

To install, I have cloned the repo and then added
```
;; You need to modify the following line
(setq load-path (cons "/path/to/lean4-mode" load-path))

(setq lean4-mode-required-packages '(dash f flycheck lsp-mode magit-section s))

(require 'package)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(package-initialize)
(let ((need-to-refresh t))
  (dolist (p lean4-mode-required-packages)
    (when (not (package-installed-p p))
      (when need-to-refresh
        (package-refresh-contents)
        (setq need-to-refresh nil))
      (package-install p))))

(require 'lean4-mode)
```

to my init file (changing the path accordingly).

**Comment by akirak** [comment]
@la-marc The file needs to be in the same directory as the elisp file. Apologies, I have to fix the instruction.

Could you please move the JSON file to the directory and try again? Thanks.

**Comment by la-marc** [comment]
I have copied the json file to the elisp files directory, but still get the same error after uncommenting the input part in `lean4-mode.el`.
I have little experience with Emacs or Elisp, but the `lean4-input.el` seems to have references to the `/data/` directory at various places. But maybe I'm misreading what happens there.

**Comment by akirak** [comment]
> the lean4-input.el seems to have references to the /data/ directory at various places. 

You are right. The following comment was my mistake. Sorry for that.

> Do you have abbreviations.json in the same directory as lean4-mode.el? 

The input method works on my computer. Have you tried `toggle-debug-on-error` before you enabling `lean4-mode` or yactivating the input method?

**Comment by la-marc** [comment]
I see, thanks for clarifying!
I tried, I'm getting
```
Debugger entered--Lisp error: (error "Can’t activate input method ‘Lean’")
  error("Can't activate input method `%s'" "Lean")
  activate-input-method("Lean")
  set-input-method("Lean")
  (let ((delay-mode-hooks t)) (prog-mode) (setq major-mode 'lean4-mode) (setq mode-name "Lean 4") (progn (if (get 'prog-mode 'mode-class) (put 'lean4-mode 'mode-class (get 'prog-mode 'mode-class))) (if (keymap-parent lean4-mode-map) nil (set-keymap-parent lean4-mode-map (current-local-map))) nil nil) (use-local-map lean4-mode-map) (set-syntax-table lean4-syntax-table) (setq local-abbrev-table lean4-abbrev-table) (set (make-local-variable 'comment-start) "--") (set (make-local-variable 'comment-start-skip) "[-/]-[ \11]*") (set (make-local-variable 'comment-end) "") (set (make-local-variable 'comment-end-skip) "[ \11]*\\(-/\\|\\s>\\)") (set (make-local-variable 'comment-padding) 1) (set (make-local-variable 'comment-use-syntax) t) (set (make-local-variable 'font-lock-defaults) lean4-font-lock-defaults) (set (make-local-variable 'indent-tabs-mode) nil) (set 'compilation-mode-font-lock-keywords 'nil) (require 'lean4-input) (set-input-method "Lean") (set (make-local-variable 'lisp-indent-function) 'common-lisp-indent-function) (lean4-set-keys) (if (fboundp 'electric-indent-local-mode) (electric-indent-local-mode -1)) (let ((--dolist-tail-- lean4-hooks-alist)) (while --dolist-tail-- (let ((x2 (car --dolist-tail--))) (progn (ignore (consp x2)) (let* ((x3 ...) (x4 ...)) (let (... ...) (add-hook hook fn nil ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (lean4-mode-setup))
  (progn (make-local-variable 'delay-mode-hooks) (let ((delay-mode-hooks t)) (prog-mode) (setq major-mode 'lean4-mode) (setq mode-name "Lean 4") (progn (if (get 'prog-mode 'mode-class) (put 'lean4-mode 'mode-class (get 'prog-mode 'mode-class))) (if (keymap-parent lean4-mode-map) nil (set-keymap-parent lean4-mode-map (current-local-map))) nil nil) (use-local-map lean4-mode-map) (set-syntax-table lean4-syntax-table) (setq local-abbrev-table lean4-abbrev-table) (set (make-local-variable 'comment-start) "--") (set (make-local-variable 'comment-start-skip) "[-/]-[ \11]*") (set (make-local-variable 'comment-end) "") (set (make-local-variable 'comment-end-skip) "[ \11]*\\(-/\\|\\s>\\)") (set (make-local-variable 'comment-padding) 1) (set (make-local-variable 'comment-use-syntax) t) (set (make-local-variable 'font-lock-defaults) lean4-font-lock-defaults) (set (make-local-variable 'indent-tabs-mode) nil) (set 'compilation-mode-font-lock-keywords 'nil) (require 'lean4-input) (set-input-method "Lean") (set (make-local-variable 'lisp-indent-function) 'common-lisp-indent-function) (lean4-set-keys) (if (fboundp 'electric-indent-local-mode) (electric-indent-local-mode -1)) (let ((--dolist-tail-- lean4-hooks-alist)) (while --dolist-tail-- (let ((x2 (car --dolist-tail--))) (progn (ignore (consp x2)) (let* (... ...) (let ... ...))) (setq --dolist-tail-- (cdr --dolist-tail--))))) (lean4-mode-setup)))
  lean4-mode()
  funcall-interactively(lean4-mode)
  #<subr call-interactively>(lean4-mode record nil)
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> lean4-mode record nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (lean4-mode record nil))
  call-interactively(lean4-mode record nil)
  command-execute(lean4-mode record)
  execute-extended-command(nil "lean4-mode" nil)
  funcall-interactively(execute-extended-command nil "lean4-mode" nil)
  #<subr call-interactively>(execute-extended-command nil nil)
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> execute-extended-command nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (execute-extended-command nil nil))
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```
but it's hard for me to tell from that what exactly goes wrong.
If you think this error is caused by some quirks in my setup/distribution and won't affect majority of the users, I can try to do without the input mode for now and try to debug myself when I have more time :).

**Comment by akirak** [comment]
@la-marc Can you evaluate the following expression and then turn on `lean4-mode` again?

```elisp
(lean4-input-setup)
```

**Comment by dwrensha** [comment]
I'm seeing the same error. According to `git bisect`, the problem started in https://github.com/leanprover/lean4-mode/commit/ca945cdb8903a306937874b2ec831b9eba548680 (i.e. #27).

**Comment by dwrensha** [comment]
```
$ emacs --version
GNU Emacs 27.1
Copyright (C) 2020 Free Software Foundation, Inc.
GNU Emacs comes with ABSOLUTELY NO WARRANTY.
You may redistribute copies of GNU Emacs
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
```

**Comment by la-marc** [comment]
> @la-marc Can you evaluate the following expression and then turn on `lean4-mode` again?
> 
> ```emacs-lisp
> (lean4-input-setup)
> ```

It does help, indeed, and almost solves the problem. After evaluating this (in `ielm`), `lean4-mode` does start without an error and "Lean" is listed among input methods (previously it was not). The lean input method does not work right after running `lean4-mode`, but started working after `M-x set-input-method` and choosing "Lean".

**Comment by akirak** [comment]
I see, the culprit is the following expression:

```elisp
(cl-eval-when (load)
  (lean4-input-setup))
```

I added `cl-eval-when` to avoid an error during byte-compilation. The expression should be evaluated only once before the input method is activated. It doesn't work for users of package.el, so it needs to be fixed.

**Comment by akirak** [comment]
I have sent a PR that should fix this issue (#31). Thanks!

**Comment by la-marc** [comment]
Great, thanks!
